platform/darwin/src/collator.mm

```diff
@@ -0,0 +1,63 @@
+#include <mbgl/style/expression/collator.hpp>
+
+#import <Foundation/Foundation.h>
+
+namespace mbgl {
+namespace style {
+namespace expression {
+    
+class Collator::Impl {
+public:
+    Impl(bool caseSensitive, bool diacriticSensitive, optional<std::string> locale_)
+    : options(caseSensitive ? NSCaseInsensitiveSearch : 0 |
+              diacriticSensitive ? NSDiacriticInsensitiveSearch : 0)
+    , locale(locale_ ?
+                [[NSLocale alloc] initWithLocaleIdentifier:[NSString stringWithUTF8String:locale_->c_str()]] :
+                [NSLocale currentLocale])
+    {}
+    
+    bool operator==(const Impl& other) const {
+        return options == other.options &&
+            [[locale localeIdentifier] isEqualToString:[other.locale localeIdentifier]];
+    }
+    
+    int compare(const std::string& lhs, const std::string& rhs) const {
+        NSString* nsLhs = [NSString stringWithUTF8String:lhs.c_str()];
+        // TODO: verify "abc" != "abcde" -- the "range" argument seems strange to me
+        // https://developer.apple.com/documentation/foundation/nsstring/1414561-compare
```

Comment:
Yes, it does seem weird that there isn’t a `-compare:options:locale:` that doesn’t take an `NSRange`. It looks like [CFString has the same quirk](https://developer.apple.com/documentation/corefoundation/1543315-cfstringcomparewithoptionsandloc?language=objc).

Reply:
We could implement this ourselves with a length check -- ie:
- First check that lhs == rhs over lhs's range
- If they're equal, check if rhs is longer than lhs, in which case rhs > lhs no matter what's in the extra characters
But I can imagine cases where that's not necessarily correct because the "extra characters" affect the sort order of the original strings (e.g. "ue" gets split up but it's in a german locale that treats "ue" as equal to "ü")

Reply:
The case below is the closest I could come to an unexpected result, but it’s pretty contrived because we’d never compare only part of the receiving string. (Pardon the Swift; the playground makes it much easier to try out possibilities.)

<img width="761" alt="strasse" src="https://user-images.githubusercontent.com/1231218/40035989-9d770c5c-57b9-11e8-9537-281dc7ed7098.png">

Reply:
I also verified the pre-fix sort behavior, and the only other concern would be character folding right at the end of the range -- however, it appears Apple doesn't do "ü"->"ue" type folding.